### ISSUE TITLE
Getting fatal error upon purchase attempt

### ISSUE THREAD

**Issue by mintplugins** [RebelCode/edd-bookings#30]
I am on PHP version 5.6.3 and getting this error when attempting to purchase a Bookings product:

```
[13-Jul-2017 14:16:36 UTC] PHP Fatal error:  Uncaught exception 'InvalidArgumentException' with message 'Argument #1 must be a numeric value or a Aventura\Diary\DateTime instance.' in /Users/philjohnston/Documents/wps/edd2.dev/wp-content/plugins/edd-bookings/vendor/rebelcode/diary/src/DateTime.php:38
Stack trace:
#0 /Users/philjohnston/Documents/wps/edd2.dev/wp-content/plugins/edd-bookings/includes/Controller/CartController.php(238): Aventura\Diary\DateTime->__construct(NULL)
#1 /Users/philjohnston/Documents/wps/edd2.dev/wp-content/plugins/edd-bookings/includes/Controller/CartController.php(197): Aventura\Edd\Bookings\Controller\CartController->validateItem(0, Array)
#2 [internal function]: Aventura\Edd\Bookings\Controller\CartController->validate()
#3 /Users/philjohnston/Documents/wps/edd2.dev/wp-includes/class-wp-hook.php(296): call_user_func_array(Array, Array)
#4 /Users/philjohnston/Documents/wps/edd2.dev/wp-includes/class-wp-hook.php(323): WP_Hook->apply_filters('', Array)
#5 /Users/philjohnston/Documents/wps/edd2.dev/wp-includes/plugin.php(453): W in /Users/philjohnston/Documents/wps/edd2.dev/wp-content/plugins/edd-bookings/vendor/rebelcode/diary/src/DateTime.php on line 38

```

**Comment by mintplugins** [comment]
This is using version 2.2.4

**Comment by pippinsplugins** [comment]
I get the same error.

I've reverted the update to 2.2.4 until this is fixed.

**Comment by mecha** [comment]
Failed to reproduce this error on PHP 5.5 and PHP 5.6.

@mintplugins @pippinsplugins Can you kindly share the configuration for the related Download?

**Comment by pippinsplugins** [comment]
It's happening to me with PHP 7.1.1

Configuration: 

<img width="1187" alt="screen shot 2017-07-13 at 10 21 11 am" src="https://user-images.githubusercontent.com/1034109/28173760-0461973e-67b5-11e7-8207-4b2f67603a3e.png">

**Comment by mecha** [comment]
I've tried those settings just now and cannot reproduce. To clarify, this error should be triggered when the Purchase button is clicked, correct?

Also, do you guys have any other extensions active? Particularly any extensions that interact with the cart session data?

----

**Tech note:**

The `DateTime` instance is receiving a `null` timestamp. The `null` value seems to be a consequence of the [filter failing to validate the integer](https://github.com/RebelCode/edd-bookings/blob/v2.2.4/includes/Controller/CartController.php#L236) timestamp that is retrieved from the cart (stored in the session).

**Comment by pippinsplugins** [comment]
The error happens when submitting checkout, not when the item is added to
the cart.

**Comment by markzahra** [comment]
What other themes and plugins do you guys have activated when this happens?

**Comment by pippinsplugins** [comment]
Vendd for the theme.

A number of EDD plugins but none that would interfere here.

This should be really easy to replicate. Both Phil and I did it first try.

1. Add a bookable product to the cart
2. Go to Checkout
3. Fill in the info
4. Click submit. The spinner will just continue to spin and never finish.
5. Check PHP error log to see error.

**Comment by mecha** [comment]
@pippinsplugins Yes, that's the process that we're using to replicate the problem. Our checkout page submits without any problems and any errors. The EDD Payment and Booking are both created with the correct information.

I've just tried testing with the Vendd theme and nothing else but EDD core and EDD Bookings active, on EDD `2.7.11`, WordPress `4.8` and using PHP versions `5.3.9`, `5.3.21`, `5.5.30`, `5.6.3` and `5.6.20`. @markzahra has also tested on PHP `5.4.19`.

This to me indicates a difference in WordPress setup, especially considering that I have already narrowed down the problem to the cart session data failing validation. It would help a lot to have a list of active EDD extensions. Even if futile, we are out of options at this point. Replication needs to happen before we can fix the problem.

**Comment by mintplugins** [comment]
I'm trying to set this up on a brand new site with no other plugins installed but I can't even activate it. Upon activation I keep getting:

```
Plugin could not be activated because it triggered a fatal error.

Fatal error: Uncaught Error: Call to undefined function Aventura\Edd\Bookings\simplexml_load_file() in /var/www/eddtests.mintplugins.com/wp-content/plugins/edd-bookings/includes/Plugin.php:746 Stack trace: #0 /var/www/eddtests.mintplugins.com/wp-content/plugins/edd-bookings/includes/Plugin.php(723): Aventura\Edd\Bookings\Plugin->loadXmlFile('/var/www/eddtes...') #1 /var/www/eddtests.mintplugins.com/wp-content/plugins/edd-bookings/integrations/settings-config.php(14): Aventura\Edd\Bookings\Plugin->loadConfigFile('settings') #2 /var/www/eddtests.mintplugins.com/wp-content/plugins/edd-bookings/includes/Plugin.php(710): include_once('/var/www/eddtes...') #3 /var/www/eddtests.mintplugins.com/wp-content/plugins/edd-bookings/edd-bookings.php(182): Aventura\Edd\Bookings\Plugin->loadDirectory('integrations') #4 /var/www/eddtests.mintplugins.com/wp-admin/includes/plugin.php(1882): include('/var/www/eddtes...') #5 /var/www/eddtests.mintplugins.com/wp-admin/plugins.php(164): plugin_sandbox_scrape('edd-bookings/ed...') #6 {main} thr in /var/www/eddtests.mintplugins.com/wp-content/plugins/edd-bookings/includes/Plugin.php on line 746
```

Can I not download the plugin from Github and install it? Very difficult to test this in its current state.

**Comment by markzahra** [comment]
Hey guys, 

Miguel has tested his current setup on PHP 7.0 again to confirm that activation works, and it did. He also tested uploading the zip file on a new WordPress site to confirm the same thing, and it worked. I also tested installing the plugin on another of my local sites. It worked, as did the checkout process.

@mintplugins from the error message you mentioned last, it is possible that you simply are missing the `libxml` extension. This should be enabled by default since 5.0. Can you please check for that?

As we've asked before, we need a list of the other extensions and plugins you have installed on your test sites. We need to rule out any possible conflicts as right now we can not reproduce this issue in any way. Have any of the customers who updated run into the same issues?

**Comment by mintplugins** [comment]
@markzahra I am setting this up on a Digital Ocean server and it does not include libxml by default. I installed PHP using their PHP7.0 package. 

This needs a separate issue so that the plugin fails gracefully if it is missing a package. I will open one for that.

**Comment by mintplugins** [comment]
I have just replicated using only these plugins active:

- EDD
- EDD Bookings

The problem seems to be coming from the incorrect purchase button being displayed - or even it showing 2 purchase buttons at times. 

See screenshot:
<img width="550" alt="screen shot 2017-07-14 at 8 29 38 am" src="https://user-images.githubusercontent.com/1638128/28212314-d0da68a4-686e-11e7-8dd5-43687e376d11.png">

If I click the right purchase button, everything works fine. 

But if I click the left purchase button, it fails as described above. 

Likely the issue would be resolved if there was only one purchase button and it was the correct one.

**Comment by markzahra** [comment]
@mintplugins you're not using the Vendd theme like Pippin are you? What theme is that please? If you can send us a zip file for it we'll test with your exact version.

**Comment by mintplugins** [comment]
It's called TwentySeventeen and it comes default with WordPress. You can likely find it in your WordPress dashboard under "Themes". If it isn't there you can search for it and it will come up as it is on WordPress's themes repo.

**Comment by mintplugins** [comment]
Just to note, this is the 3rd site in a row we've replicated it on. 2 were local installations and this one is hosted on Digital Ocean on PHP7.0

**Comment by markzahra** [comment]
Thank you @mintplugins. Did you test to see if you get the same duplicate buttons issue on other themes? 

We've also set up another site locally and could not replicate the issue.

I've also set up a live site at test.eddbookings.com. I installed 2.2.4 from scratch, no trouble there. I used the TwentySeventeen theme, and I had no duplicate buttons either. Check it out:

http://test.eddbookings.com/downloads/consultation/

I then went through the same process as you guys went through to set up another download with bookings and tried to purchase it. It worked fine. You can see and try it for yourselves again:

http://test.eddbookings.com/downloads/gh-30/

Here are temporary login details to that site. Please compare to your setups and let us know if anything at all is different.

U:  edd2017
P:  4Sq(IRyW0qy5YGKjE3*9c(tX

**Comment by markzahra** [comment]
By the way, you mentioned that "this is the 3rd site in a row we've replicated it on". Is that referring only to the checkout fatal error, or the installation fatal error, or the duplicate purchase button? Which sites produced which issues?

**Comment by mintplugins** [comment]
@markzahra This is referring to the checkout fatal error - the one this issue was opened for. The duplicate purchase button happens intermittently but I wanted to make sure you knew about it so you could keep an eye on it. 

I am also getting the duplicate purchase button intermittently on Vendd, yes. 

One thing that may have been different on a couple of our local sites is that FES was active and some of the bookings products were submitted by Vendors. Give that a shot to see.

**Comment by mecha** [comment]
@mintplugins Myself and @markzahra have set up FES on the [test site](http://test.eddbookings.com). I created a Download with bookings enabled through FES as a vendor and checkout goes through without failure.

Unless we see the problem in action we are not going to be able to provide a fix. Can you try it out on our test site and see if you can replicate it here? Perhaps also check if there's anything relevant that is different in our setup?

**Comment by markzahra** [comment]
Also, have you guys only seen this issue on your own setups or was this reported by a customer? If we can get access to one of your sites (or that of a customer) where the issue is presenting itself it would help a lot.

**Comment by mintplugins** [comment]
I did check it on your test site but it's not happening there. I can't seem to see any differences. Not sure what to say. On 2 of mine and Pippin's it is happening. On your test site it is not.

**Comment by mintplugins** [comment]
Anything you can think of that would cause it based on the error message? 

Any reason there would be two buttons?

**Comment by mecha** [comment]
What we know so far is that the error is caused by missing information in the cart. It is being triggered on checkout when the cart item data is validated and used to create a Booking. We also know that this bug was most likely not introduced in 2.2.4, simply based on the cause and the changeset for 2.2.4.

At this time we're not sure if the cart data is just missing or if it's being cleared by something else (like another plugin) before our extension has a chance to read it.

The duplicate button issue _might_ explain the missing data. Our replacement purchase button performs availability checks and adds extra information to the item that was added to the cart. The original (or duplicate) button does not. As for why the buttons both appear at the same time, we do not know.

If you have a site that is suffering from this bug and to which you can provide access, it would help tremendously.

**Comment by pippinsplugins** [comment]
I'll test again on a new test site to see what behavior I get.

**Comment by mintplugins** [comment]
I just tested this on version 2.2.3 and I am still getting the same error. When it happens, I'm pretty sure it is showing the incorrect add-to-cart button.

**Comment by markzahra** [comment]
Ok so we know it's not an issue introduced in 2.2.4 then, thanks @mintplugins. Did you test on a local site again? Is it possible to transfer that local site to a live domain so we can view it ourselves? We still cannot reproduce here.

**Comment by mintplugins** [comment]
I don't have any hosting that I can give you the SSH password for. 

Do you have a server set up that I could install it for you on?

**Comment by mintplugins** [comment]
P.S. if you want to see it in action you can go here:
https://eddtests.mintplugins.com/downloads/dog-walking/

**Comment by markzahra** [comment]
@mintplugins @pippinsplugins 

We asked for a list of the plugins you are using a number of times, and for good reason. Miguel found that you are using EDD 2.8 BETA 1 on the eddtests site sent by Phil above. We had to find this out via the JS on that site. This information could have saved us a few days. Pippin are you using this BETA version too? Fyi, we were always testing on the latest stable version, 2.7.11, which would explain why this wasn't reported by a user the moment it went out.

When we both tested using EDD 2.8 BETA 1, we both reproduced the issue of the duplicate button and the fatal error at checkout. We could reproduce the issues again using EDD 2.8 BETA 2.

Miguel will be sending a detailed report shortly.

**Comment by mecha** [comment]
When EDD Bookings is initializing the elements that it needs from EDD Core, it targets the `form` element and the Add to Cart button.

There are 2 add to cart buttons:

1. `a.edd-add-to-cart.edd-has-js`
2. `input.edd-add-to-cart.edd-no-js`

EDD Bookings targets the *visible* button. This is because of EDD having 2 buttons, one with the `edd-has-js` class, the other with `edd-no-js` class.

On EDD 2.7.11, the first button is the visible button. EDD BK attaches the events to it and hides it. It will be then shown when a date and time are selected.

On EDD 2.8 Beta 1 and 2, the second button is the visible button. EDD BK attaches the events to it and hides It will be then shown when a date and time are selected. However, in this case EDD Core also makes the first button become visible. This button is not handled by EDD BK, so no booking information is submitted when the button is clicked. This is also why 2 buttons can be seen after a date and time have been selected.

A fix from our end would be to stop catering for the second button, and always use the first; i.e. `a.edd-add-to-cart.edd-has-js`, while also force hiding the second.

**Comment by pippinsplugins** [comment]
@markzahra sorry for the delay. I've been in the middle of moving houses and have been mostly disconnected. Since it was a few days ago, I can't say for sure if I was testing on the 2.8 beta, but I'm going to test again now to be sure.

@mecha You should be able to safely only target the first one, `a.edd-add-to-cart.edd-has-js`. The input field is a legacy button that used to be used when javascript didn't work on the site. While the button is still there, we'll likely remove it in the future.

**Comment by pippinsplugins** [comment]
Definitely looks like something in 2.8 is breaking Bookings. I'm seeing this JS error on the product page with the add to cart button:

>>>Uncaught TypeError: Cannot read property 'click.eddAddToCart' of undefined
https://edd.com/vendd/wp-content/plugins/edd-bookings/assets/js/service-frontend.js?ver=2.1.2-RC1 line 760

**Comment by pippinsplugins** [comment]
I've confirmed what @mecha described above.

On EDD pre-2.8-beta versions, the `a.edd-add-to-cart` element is shown. On 2.8-beta, the `input.edd-add-to-cart` element is shown.
<img width="1116" alt="screen shot 2017-07-19 at 8 57 05 am" src="https://user-images.githubusercontent.com/1034109/28370786-c9eb993c-6c60-11e7-8926-7cc922509111.png">
<img width="1117" alt="screen shot 2017-07-19 at 8 59 14 am" src="https://user-images.githubusercontent.com/1034109/28370787-c9edb654-6c60-11e7-9117-838c6b870f98.png">

Working on tracking it down.

**Comment by pippinsplugins** [comment]
@mecha Looks like the problem only happens when Bookings is activated. The `a.edd-add-to-cart` button is the first visible button on EDD 2.8-beta and prior versions when Bookings is deactivated. The `input.edd-add-to-cart` element is somehow made visible by Bookings.

**Comment by mecha** [comment]
@pippinsplugins That makes sense and aligns with what I've found. However, very early on (on document ready and prior to Core's scripts) the `input.edd-add-to-cart` is visible, which causes Bookings to save a reference and attach events to it.

Immediately after Bookings finishes this processing, both buttons are hidden by Core ([code here](https://github.com/easydigitaldownloads/easy-digital-downloads/blob/release/2.8/assets/js/edd-ajax.js#L5)) and then the `a.edd-add-to-cart` is shown ([code here](https://github.com/easydigitaldownloads/easy-digital-downloads/blob/release/2.8/assets/js/edd-ajax.js#L6)).

However, Bookings would have already committed itself to the `input` button, and would later re-show it when the date and time are selected.

**Comment by mecha** [comment]
So in conclusion, the recommended solution is to stop trying to detect which button will be used, treat the `input` button as deprecated and always attach events to the `a` button. @pippinsplugins do you agree?

**Comment by mecha** [comment]
This is also the cause for #29

**Comment by pippinsplugins** [comment]
>>>So in conclusion, the recommended solution is to stop trying to detect which button will be used, treat the input button as deprecated and always attach events to the a button. @pippinsplugins do you agree?

Yes I agree

**Comment by mecha** [comment]
In case anyone wants to give it a test, the fix has been pushed to the [`hotfix/2.2.4`](https://github.com/RebelCode/edd-bookings/tree/hotfix/2.2.4) branch.
